apps/antidote/src/meta_data_notif_server.erl

```diff
@@ -0,0 +1,336 @@
+notify_subscribers(Snapshot, State) ->
+    Snaps0 = vector_orddict:to_list(State#state.snaps),
+    SnapsList0 = lists:reverse(Snaps0), %% Small -> Large
+    case notify_subscribers(Snapshot, _Left = [], SnapsList0, []) of
+        [] ->
+            State#state{snaps = vector_orddict:new(), min_snapshot = undefined};
+        SnapsList1 ->
+            MinSnaps = calculate_min(SnapsList1),
+            Snaps1 = vector_orddict:from_list( lists:reverse(SnapsList1) ),
+            State#state{snaps = Snaps1, min_snapshot = MinSnaps}
+    end.
+
+%% List is expected to be sorted in ascending order
+-spec calculate_min([{vectorclock:vectorclock(), term()}]) -> vectorclock:vectorclock().
+calculate_min([{ Min, _} | Rest]) ->
+    calculate_min(Min, Rest).
+
+calculate_min(Min, []) ->
+    Min;
+calculate_min(Min, [{H, _} | Rest]) ->
+    case vectorclock:ge(H, Min) of
+        true ->
+            Min;
+        false ->
+            calculate_min(vectorclock:min2(Min, H), Rest)
+    end.
+
+notify_subscribers(_Snapshot, Concurrent, [], Notifications) ->
+    send_snapshot_notifications(lists:reverse(Notifications)),
+    lists:reverse(Concurrent);
+notify_subscribers(Snapshot, Concurrent, [{SN, Clients} = R | MaybeGreater] = Snaps,
+                  Notifications) ->
+    case vectorclock:ge(Snapshot, SN) of
+        true ->
+            notify_subscribers(Snapshot, Concurrent, MaybeGreater,
+                               [{SN, Clients} | Notifications]);
+        false ->
+            case vectorclock:ge(SN, Snapshot) of
+                true ->
+                    send_snapshot_notifications(lists:reverse(Notifications)),
+                    lists:reverse(Concurrent) ++ Snaps;
+                false ->
+                    notify_subscribers(Snapshot, [R | Concurrent], MaybeGreater,
+                                       Notifications)
+            end
+    end.
+
+send_snapshot_notifications([{SN, Clients} | Rest]) ->
+    _ = [Client ! #md_stable_snapshot{sn = SN} || Client <- Clients],
+    send_snapshot_notifications(Rest);
+send_snapshot_notifications([]) ->
+    ok.
+
+add_snapshot(Snapshot, Pid, State) ->
+    #state{mons = Mons0, snaps = Snaps0, min_snapshot = MinSnap0} = State,
+    Mons1 = add_subscriber(Pid, Mons0),
+
+    {MinSnap1, Snaps1} = add_snapshot(Pid, Snapshot, MinSnap0, Snaps0),
+
+    State#state{mons = Mons1, snaps = Snaps1, min_snapshot = MinSnap1}.
+
+add_subscriber(Pid, Mons0) ->
+    case lists:keyfind(Pid, 1, Mons0) of
+        {Pid, _} -> Mons0;
+        false ->
+            MonRef = erlang:monitor(process, Pid),
+            [{Pid, MonRef} | Mons0]
+    end.
+
+-spec add_snapshot(pid(), antidote:snapshot_time(), antidote:snapshot_time(),
+                   vector_orddict:vector_orddict([pid()])) ->
+          { antidote:snapshot_time(), vector_orddict:vector_orddict([pid()]) }.
+add_snapshot(Pid, Snapshot, MinSnap, Snaps0) ->
+    Snaps1 = vector_orddict:append(Snapshot, Pid, Snaps0),
```

Comment:
How large can this vector get?

Reply:
Well main consumer of this feature was intended to be wal streaming api. For a single partition this orddict is not going to be large, but for scenarios with multi-partitioning when some partitions experience replication problems and some don't, this orddict would start accumulating all non-fully replicated transaction snapshots, which could be large.

For other users of this API, it make sense to consider on the case by case basis.

Reply:
Then I would add some form of warning into the documentation of the function that this is a potential space leak, just in case somebody runs into the issue.